lib/puppet/node/environment.rb

```diff
@@ -591,10 +591,12 @@ def perform_initial_import
       if file == NO_MANIFEST
         empty_parse_result
       elsif File.directory?(file)
-        parse_results = Puppet::FileSystem::PathPattern.absolute(File.join(file, '**/*.pp')).glob.sort.map do | file_to_parse |
-          parser.file = file_to_parse
-          parser.parse
-          end
+        # JRuby does not properly perform Dir.glob operations with wildcards,
+        # see PUP-11788 and https://github.com/jruby/jruby/issues/7836
+        parse_results = Puppet::FileSystem::PathPattern.absolute(File.join(file, '**')).glob.select {|globbed_file| globbed_file.end_with?('.pp')}.sort.map do | file_to_parse |
+                          parser.file = file_to_parse
+                          parser.parse
+                        end
```

Comment:
Could you add an rspec test to `spec/unit/file_system/path_pattern_spec` to test globbing of `**/*.pp` and mark it as pending when running under JRuby? That way we know to remove our workaround if/when JRuby fixes the issue.